docs/tutorial/fle.txt

```diff
@@ -0,0 +1,164 @@
+
+   $collection->insertOne(['encryptedField' => '123456789']);
+
+   var_dump($collection->findOne([]));
+
+
+Specifying an explicit schema for encryption
+--------------------------------------------
+
+The following example shows how to create a new key and store it in the key
+vault collection. The encrypted client configures an explicit schema for
+encryption using the newly created key.
+
+.. note::
+
+   Supplying a ```schemaMap``` provides more security than relying on JSON
+   schemas obtained from the server. It protects against a malicious server
+   advertising a false JSON schema, which could trick the client into sending
+   unencrypted data that should be encrypted.
+
+.. code-block:: php
+
+   <?php
+
+   use MongoDB\BSON\Binary;
+   use MongoDB\Client;
+   use MongoDB\Driver\ClientEncryption;
+
+   $localKey = new Binary('<binary key data>', Binary::TYPE_GENERIC);
+
+   $clientEncryptionOpts = [
+       'keyVaultNamespace' => 'admin.datakeys',
+       'kmsProviders' => [
+           'local' => ['key' => $localKey],
+       ],
+   ];
+
+   $client = new Client();
+   $clientEncryption = $client->createClientEncryption($clientEncryptionOpts);
+
+   // Create new key in the key vault and store its ID for later use
+   $keyId = $clientEncryption->createDataKey('local');
+
+   $autoEncryptionOpts = [
+       'keyVaultNamespace' => 'admin.datakeys',
+       'kmsProviders' => [
+           'local' => ['key' => $localKey],
+       ],
+       'schemaMap' => [
+           'test.coll' => [
+               'bsonType' => 'object',
+               'properties' => [
+                   'encryptedField' => [
+                       'encrypt' => [
+                           'keyId' => [$keyId],
+                           'bsonType' => 'string',
+                           'algorithm' => ClientEncryption::AEAD_AES_256_CBC_HMAC_SHA_512_DETERMINISTIC,
+                       ],
+                   ],
+               ],
+           ],
+       ],
+   ];
+
+   $encryptedClient = new Client(null, [], ['autoEncryption' => $autoEncryptionOpts]);
+
+   $collection = $encryptedClient->selectCollection('test', 'coll');
+   $collection->drop(); // clear old data
+
+   $collection->insertOne(['encryptedField' => '123456789']);
+
+   var_dump($collection->findOne([]));
+
+
+Manually encrypting and decrypting values
+-----------------------------------------
+
+In the Community edition of MongoDB, you will have to manually encrypt and
+decrypt values before storing them in the database. The following example
+assumes that you have already created an encryption key in the key vault
+collection and explicitly encrypts and decrypts values in the document.
+
+.. code-block:: php
+
+   <?php
+
+   use MongoDB\BSON\Binary;
+   use MongoDB\Client;
+   use MongoDB\Driver\ClientEncryption;
+
+   $localKey = new Binary('<binary key data>', Binary::TYPE_GENERIC);
+   $keyId    = new Binary('<UUID of encryption key', Binary::TYPE_UUID);
```

Comment:
Should we add a comment above this to make the connection to `ClientEncryption::createDataKey()`? I believe we only show that being called in the previous `schemaMap` example, which is also for the enterprise, automatic encryption. That all seems a bit odd, as ClientEncryption seems like it's only useful for the explicit encryption but we end up using `createDataKey` alone for automatic encryption as well.

Reply:
I've changed the example and added creating an encryption key using `createDataKey`, so there is no confusion as to where this is coming from. However, this might lead users to believe that they should create a new encryption key anytime they want to encrypt data, which might not be desirable. We may want to address this in the tutorial if we indeed don't recommend doing this.

Reply:
@kevinAlbs I assume we don't expect users to create new data keys each time they need to invoke encryption. Do we expect them to use this method once and then either (a) directly query the key vault collection for future use or (b) save the ID and then hard-code it into their application/environment? I read through most of [ClientEncryption](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#clientencryption) and related references to `keyId` but I don't see much elaboration on this.

Reply:
> I assume we don't expect users to create new data keys each time they need to invoke encryption.

Yes, but they may be creating keys on the fly as part of their application. E.g. if they're creating a per-user data key.

> Do we expect them to use this method once and then either (a) directly query the key vault collection for future use or (b) save the ID and then hard-code it into their application/environment?

That is one way. But they may also associate one or more keyAltNames based on other data (e.g. a user identifier) so would not need to query or hard code the ID.

Reply:
alternative name is a good point. I've added an example to show how this is used and also explained that encryption keys shouldn't be created every time a value is encrypted. I hope this clarifies the solution.